### DISCUSSION TITLE
Project #5: AI/ML in diffusion MRI processing

### DISCUSSION THREAD

**Discussion by itellaetxe** [dipy/dipy#3072] [#GSoC/GSoD]
Dear DIPY mentors,
(@skoudoro @pjsjongsung @sreekarchigurupati)

## Introduction
I am Iñigo Tellaetxe Elorriaga, BSc and MSc in Biomedical Engineering, and first year PhD student in the Computational Neuroimaging Laboratory in IIS Biobizkaia, in the Basque Country. I am interested in working on [Project #5: Project ideas using AI/ML in Diffusion MRI processing](https://github.com/dipy/dipy/wiki/Google-Summer-of-Code-2024#project-5-project-ideas-using-aiml-in-diffusion-mri-processing).

## Relevant Experience
I have considerable experience handling multiple neuroimaging modalities, especially structural and diffusion MRI, as I work with these daily. My thesis is focused on Brain Ageing from a computational perspective (under the supervision of [Asier Erramuzpe](https://github.com/erramuzpe) and [Jesús Cortés](https://jesuscortes.info/)), so I am familiar with neuroimaging preprocessing tools and pipelines such as nibabel, DIPY (I attended last year's workshop), MRIQC, nilearn, FreeSurfer, FSL, ANTs, and fMRIprep.

I previously worked in an image-to-image conversion problem to generate synthetic CT from MRI, to reduce ionizing radiation exposure in spine surgery patients in my MSc Thesis. In addition, I got the opportunity to study structural connectivity features in Alzheimer's Disease, using ADNI's data across different clinical categories, identifying key connectivity alterations for developing AD.

Regarding Open Source experience, in our lab we developed a Python package for Age Modelling using Machine Learning called [ageml](https://github.com/compneurobilbao/ageml). Thus, I believe I have a good grasp on how the Open Source development process works, combined with strong Python skills.

## Proposal
I am eager to contribute to this project from virtually any proposed perspective, but I have a couple of ideas from my side:
- Optimizing any step of standard dMRI preprocessing steps using DL techniques (Eddy correction, susceptibility correction, diffusion model computation, etc.).
- Correcting/Generating dMRI and tractography data conditioned on clinical characteristics (demographic and pathological)

I would really love to contribute to DIPY because I have been using it for some years and I finally found the opportunity to do it. I truly believe I can make a meaningful contribution to it. I would like to discuss these ideas with you to see if we can find an interesting useful project. In the meantime, I will handle some open issues so I can get familiar with contributing to DIPY.

Thanks,

Iñigo

**Comment by pjsjongsung** [comment]
Hello and thank you for your interest in the project. Both of your proposal sounds great, and it would be nice to see in detail what ideas you have later when you apply. For example, which step in the preprocessing pipeline are you most interested? You would have to do some literature search on the topic as well.

We are looking forward to reading your full proposal!

**Comment by skoudoro** [comment]
Hi @itellaetxe,

Thank you for you first contribution https://github.com/dipy/dipy/pull/3086.

Now, the next step is to prepare your GSoC proposal. When have a draft almost ready, please share it to us privately that we can review it before submission to gsoc platform. 

We usually do 1 review per candidate who made their first contribution. it helps to avoid basic mistake. 

Good luck

**Reply by itellaetxe** [reply]
Excellent, @skoudoro. I hope we can resolve the doubts concerning the contribution soon!

I will work on preparing the proposal. Thank you!

**Comment by itellaetxe** [comment]
Dear @skoudoro and @pjsjongsung 

I have been reading some literature along the lines of tractography filtering ([paper](https://doi.org/10.1016/j.media.2021.102126)) and generative tractography sampling ([paper](http://arxiv.org/abs/2204.10891)), from @jhlegarreta, as well as other related works. I think DIPY could benefit from the methods described in these works and that they could possibly be extended somehow. 

As I mentioned in the first post, generating tractography data conditioned on demographic features such as age and/or clinical characteristics (e.g. mildly cognitively impaired patients, dementia, AD) could be interesting both in the methodological and clinical aspects of it. Regarding the methodology, because it would enable sampling tractograms somehow flexibly, to get "normative" examples of desired populations, for tractometry/connectomics studies. Also, making inference with Autoencoder-like networks is not extremely expensive computationally. Finding a way to condense the learned latent information into "template" tractograms would be interesting, similar to what the MNI template means in the field. What do you think about this? Could it be on the lines of DIPY?

I am writing this because I don't want to write a proposal on a topic that perhaps is not interesting to the DIPY community, so I wanted to check before delving into it more deeply. If you have other directions you would like me to direct my efforts to, propose them, please, and I will make my best coming up with an interesting idea.

**Reply by pjsjongsung** [reply]
The topic sounds good! I am personally interested in tractography and it is a great direction DIPY can push.

I am a little unsure of how we would find such tractography data with demographic features and whether the method is at this stage generalizable enough for implementation in DIPY. But I think Jon would know best about this subject. @jhlegarreta do you have any thoughts about this project?

**Reply by skoudoro** [reply]
To me, it is going to the right direction. Also, let us know @jhlegarreta if you have time to be an additional mentor this year concerning this topic (ps: no need to attend all gsoc meeting)

**Reply by jhlegarreta** [reply]
Hi all,
thanks for tagging me.

Probably such data exists and is openly available, at least to some extent.

The core of the project proposal seems to be coming up with a well-funded hypothesis about the model that would best support creating such a template or predictor using an AE.

Note that my implementation is in PyTorch; re-implementing the autoencoder model (before other additions related to the project) it in Tensorflow (if required) would not be the hardest part probably.

Happy to talk over a call as necessary.

**Reply by pjsjongsung** [reply]
Thank you @jhlegarreta . @itellaetxe so we are more than happy to work with you on this project. If you need help with the topic we will be more than happy to discuss this further with you. We'll be waiting for your proposal!

**Reply by itellaetxe** [reply]
Thank you for considering my idea @skoudoro @pjsjongsung. Also thank you @jhlegarreta for the valuable input. I really would like to talk over a call about the idea, I will send you an email.

Knowing that this is interesting for DIPY, I will further develop the idea and write the proposal soon. If I have any more doubts I will let you know. I will keep you updated!

**Comment by itellaetxe** [comment]
Hi again @skoudoro @pjsjongsung @jhlegarreta ,

For the proposal, I think I have a couple of interesting ideas, after reading some literature suggested by Jon Haitz (thanks again for the info!).
My project would be named something like: "Conditional Tractogram Generation Based on Age and neurodegenerative clinical categories".

This would be an early summary of the proposal. Let me know your opinions on it!

### Motivation
Generating tractography data conditioned on demographic features such as age or neurodegenerative disease status as a means to provide normative or populational information of structural brain connectivity from a learned interpretable latent space.

### Main objective
To integrate in DIPY a deep learning model capable of producing synthetic whole-brain tractograms, conditioned on a specified age and clinical category (e.g.: Healthy VS. Alzheimer's Disease).

### Secondary objectives
(steps to accomplish the main one)
- To integrate unconditioned synthetic tractogram generation capabilities in DIPY using @jhlegarreta's AutoEncoder ([ref](http://arxiv.org/abs/2204.10891)) by translating his PyTorch model to Tensorflow, as the main dependency for DL in DIPY is Tensorflow. Make a function to get the tractograms,`generate_synth_tractogram()`.
- To investigate a combination of attribute-regularized variational autoencoders ([ref](http://arxiv.org/abs/2004.05485)) and domain invariant variational autoencoders ([ref](http://arxiv.org/abs/1905.10427)) for generating whole tractograms conditioned on continuous (age) and discrete (clinical category) input variables at the same time. Run experiments on publicly available data.
- Train the model with public data (ADNI3) and integrate it in DIPY in a function `generate_conditioned_synth_tractogram(age: int, class: str)`. `class` being either `"healthy"` or `"ad"` for Alzheimer's Disease.

### Benefits to the community
Enabling data generation for:
- Tractometry benchmarks
- Clinical studies (probably related to ageing) that need a synthetic subject template with the desired age and disease status.
- Synthetic datasets

### Deliverables
- Report on the State of the Art review 
- Function, tests, and documentation for generating tractograms with Jon Haitz's AutoEncoder. (Translated from PyTorch to Tensorflow)
- Report of Training results of the proposed Variational Autoencoder architecture.
- Function, tests, and documentation for generating conditioned tractograms with the proposed architecture. (Directly in Tensorflow)

**Reply by jhlegarreta** [reply]
Looks good to me. Not sure the about the quality of the VAE vs AE reconstruction; might be worthwhile starting with the AE, then checking if the VAE reconstruction is good or comparable, and moving forward based on those results.

**Reply by skoudoro** [reply]
Hi @jhlegarreta, we need to register mentors. Please, contact me if you can do it and do not hesitate to share your restriction and how much time you can commit. Since your are not the main mentor, it should be fine if you do not have a lot of time (however mentor registration is important).
Thank you for the feedback

**Reply by itellaetxe** [reply]
Dear @skoudoro @pjsjongsung @jhlegarreta,

I just finished the first version of my proposal. I invited you to the Google Document containing it, you should have a notification in your mails. I think it is ready for you to review, so please suggest any changes you see appropriate.

@pjsjongsung I could not find your email address, so I sent the link of the document to Serge and Jon Haitz, so they can forward it to you :)

Thank you!

**Reply by jhlegarreta** [reply]
Thanks. Left some comments. Shared the document with Jong Sung.

**Reply by pjsjongsung** [reply]
Thanks @jhlegarreta . I left a few comments as well.

**Reply by jhlegarreta** [reply]
Had another pass to it. Resolved some comments, made minor suggestions. After those are done, it is ready to go on my end.

**Reply by itellaetxe** [reply]
Thank you both for the comments and the suggestions. I will send the document this Friday to let the doc breathe a bit, just in case I missed something. The deadline is April 2, 18:00 UTC.
